plugin/db/mongodb/migrate.go

```diff
@@ -85,16 +115,320 @@ func (driver *Driver) SetupMigrationIfNeeded(ctx context.Context) error {
+	if err := cursor.Err(); err != nil {
+		return nil, errors.Wrapf(err, "failed to find largest version since baseline")
+	}
+	return nil, nil
+}
+
+// InsertPendingHistory will insert the migration record with pending status and return the inserted ID.
+func (driver *Driver) InsertPendingHistory(ctx context.Context, _ *sql.Tx, sequence int, _ string, m *db.MigrationInfo, storedVersion, statement string) (insertedID int64, err error) {
+	collection := driver.client.Database(migrationHistoryDefaultDatabase).Collection(migrationHistoryDefaultCollection)
+
+	retryTimes := 3
+	nextID := int64(0)
+	for i := 0; i < retryTimes; i++ {
+		currentTimestamp := getMongoTimestamp()
+		nextID, err = driver.getMigrationHistoryNextID(ctx)
+		if err != nil {
+			return 0, errors.Wrapf(err, "failed to get next migration history ID")
+		}
+
+		document := bson.M{
+			"id":                    nextID,
+			"created_by":            m.Creator,
+			"created_ts":            currentTimestamp,
+			"updated_by":            m.Creator,
+			"updated_ts":            currentTimestamp,
+			"release_version":       m.ReleaseVersion,
+			"namespace":             m.Namespace,
+			"sequence":              int64(sequence),
+			"source":                m.Source,
+			"type":                  m.Type,
+			"status":                db.Pending,
+			"version":               storedVersion,
+			"description":           m.Description,
+			"statement":             statement,
+			"execution_duration_ns": int64(0),
+			"issue_id":              m.IssueID,
+			"payload":               m.Payload,
+		}
+		if _, err = collection.InsertOne(ctx, document); err != nil {
+			// Detect if it's a duplicate key error.
+			if driverErr, ok := err.(mongo.WriteException); ok {
+				find := false
+				for _, writeErr := range driverErr.WriteErrors {
+					if writeErr.Code == 121 {
+						log.Info("Duplicate key error with migration_history id: %d, retrying...", zap.Int64("id", nextID))
+						find = true
+						break
+					}
+				}
+				if find {
+					continue
+				}
+			}
+			return 0, errors.Wrapf(err, "failed to insert a pending migration history record")
+		}
+		return nextID, nil
+	}
+	return nextID, nil
+}
+
+// UpdateHistoryAsDone will update the migration record as done.
+func (driver *Driver) UpdateHistoryAsDone(ctx context.Context, _ *sql.Tx, migrationDurationNs int64, _ string, insertedID int64) error {
+	collection := driver.client.Database(migrationHistoryDefaultDatabase).Collection(migrationHistoryDefaultCollection)
+	filter := bson.M{
+		"id": insertedID,
+	}
+	update := bson.M{
+		"$set": bson.M{
+			"status":                db.Done,
+			"execution_duration_ns": migrationDurationNs,
+		},
+	}
+
+	updateResult, err := collection.UpdateOne(ctx, filter, update)
+	if err != nil {
+		return errors.Wrapf(err, "failed to update a migration history as done")
+	}
+	if updateResult.MatchedCount == 0 {
+		return errors.Errorf("failed to find migration history with ID %d", insertedID)
```

Comment:
Need to provide more context, otherwise can't differentiate from UpdateHistoryAsFailed at line 384

Reply:
IMO this is the caller's responsibility to add more context, like using `errors.wrap(err, "failed to UpdateHistoryAsFailed")`.
But providing more context in callee is ok for me.